In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datetime as dt

%matplotlib inline
sns.set_style('darkgrid')
sns.set_palette('viridis')

In [2]:
spray = pd.read_csv('../data/spray.csv')

In [3]:
spray.head()

Date        Time   Latitude  Longitude
0  2011-08-29  6:56:58 PM  42.391623 -88.089163
1  2011-08-29  6:57:08 PM  42.391348 -88.089163
2  2011-08-29  6:57:18 PM  42.391022 -88.089157
3  2011-08-29  6:57:28 PM  42.390637 -88.089158
4  2011-08-29  6:57:38 PM  42.390410 -88.088858

In [4]:
spray.isnull().sum()

Date           0
Time         584
Latitude       0
Longitude      0
dtype: int64

In [5]:
spray.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14835 entries, 0 to 14834
Data columns (total 4 columns):
Date         14835 non-null object
Time         14251 non-null object
Latitude     14835 non-null float64
Longitude    14835 non-null float64
dtypes: float64(2), object(2)
memory usage: 463.7+ KB


In [6]:
spray[spray['Time'].isnull()]

Date Time   Latitude  Longitude
1030  2011-09-07  NaN  41.987092 -87.794286
1031  2011-09-07  NaN  41.987620 -87.794382
1032  2011-09-07  NaN  41.988004 -87.794574
1033  2011-09-07  NaN  41.988292 -87.795486
1034  2011-09-07  NaN  41.988100 -87.796014
1035  2011-09-07  NaN  41.986372 -87.794862
1036  2011-09-07  NaN  41.986228 -87.795582
1037  2011-09-07  NaN  41.984836 -87.793998
1038  2011-09-07  NaN  41.984836 -87.794670
1039  2011-09-07  NaN  41.984884 -87.795198
1040  2011-09-07  NaN  41.982820 -87.793182
1041  2011-09-07  NaN  41.982772 -87.793950
1042  2011-09-07  NaN  41.982772 -87.795006
1043  2011-09-07  NaN  41.983444 -87.795246
1044  2011-09-07  NaN  41.984068 -87.795438
1045  2011-09-07  NaN  41.984596 -87.795534
1046  2011-09-07  NaN  41.984980 -87.795630
1047  2011-09-07  NaN  41.986708 -87.796110
1048  2011-09-07  NaN  41.987524 -87.796302
1049  2011-09-07  NaN  41.988820 -87.796638
1050  2011-09-07  NaN  41.989444 -87.796830
1051  2011-09-07  NaN  41.990068 -87.796974
1052  2011-09-07  NaN  41.988244 -87.797022
1053  2011-09-07  NaN  41.988196 -87.797742
1054  2011-09-07  NaN  41.988196 -87.798366
1055  2011-09-07  NaN  41.988196 -87.798942
1056  2011-09-07  NaN  41.988196 -87.799422
1057  2011-09-07  NaN  41.986372 -87.796494
1058  2011-09-07  NaN  41.986372 -87.797310
1059  2011-09-07  NaN  41.986372 -87.798030
...          ...  ...        ...        ...
1584  2011-09-07  NaN  41.992372 -87.816367
1585  2011-09-07  NaN  41.992468 -87.815839
1586  2011-09-07  NaN  41.992516 -87.815311
1587  2011-09-07  NaN  41.992564 -87.814735
1588  2011-09-07  NaN  41.992612 -87.814303
1589  2011-09-07  NaN  41.992612 -87.813775
1590  2011-09-07  NaN  41.992612 -87.813343
1591  2011-09-07  NaN  41.992660 -87.812767
1592  2011-09-07  NaN  41.992660 -87.812191
1593  2011-09-07  NaN  41.992612 -87.811663
1594  2011-09-07  NaN  41.992660 -87.811279
1595  2011-09-07  NaN  41.992660 -87.810847
1596  2011-09-07  NaN  41.992612 -87.810415
1597  2011-09-07  NaN  41.992660 -87.809455
1598  2011-09-07  NaN  41.992660 -87.809983
1599  2011-09-07  NaN  41.992708 -87.808927
1600  2011-09-07  NaN  41.992708 -87.808399
1601  2011-09-07  NaN  41.992756 -87.807775
1602  2011-09-07  NaN  41.992756 -87.807199
1603  2011-09-07  NaN  41.996164 -87.813727
1604  2011-09-07  NaN  41.995732 -87.813679
1605  2011-09-07  NaN  41.994916 -87.813727
1606  2011-09-07  NaN  41.992948 -87.813775
1607  2011-09-07  NaN  41.996644 -87.811567
1608  2011-09-07  NaN  41.996260 -87.811567
1609  2011-09-07  NaN  41.995876 -87.811615
1610  2011-09-07  NaN  41.995972 -87.810271
1611  2011-09-07  NaN  41.995684 -87.810319
1612  2011-09-07  NaN  41.994724 -87.810415
1613  2011-09-07  NaN  41.993092 -87.810415

[584 rows x 4 columns]

In [7]:
spray.iloc[11499, :]

Date          2013-08-22
Time         10:53:16 PM
Latitude          41.724
Longitude       -87.6222
Name: 11499, dtype: object

In [8]:
spray[spray['Longitude']== -87.615892]

Empty DataFrame
Columns: [Date, Time, Latitude, Longitude]
Index: []

In [9]:
order_groups = spray.groupby(['Latitude', 'Longitude'], as_index=False).count().sort_values('Date', ascending=False)

In [10]:
order_groups

Latitude  Longitude  Date  Time
11853  41.986460 -87.794225   541   541
11499  41.983917 -87.793088     2     2
0      41.713925 -87.615892     1     1
9533   41.959113 -87.719752     1     1
9522   41.959028 -87.728890     1     1
9523   41.959052 -87.725095     1     1
9524   41.959055 -87.724518     1     1
9525   41.959055 -87.711887     1     1
9526   41.959070 -87.715542     1     1
9527   41.959080 -87.737005     1     1
9528   41.959085 -87.711258     1     1
9529   41.959090 -87.716798     1     1
9530   41.959105 -87.710555     1     1
9531   41.959107 -87.722885     1     1
9532   41.959112 -87.712832     1     1
9534   41.959115 -87.710088     1     1
9520   41.959023 -87.729487     1     1
9535   41.959123 -87.709648     1     1
9536   41.959123 -87.708298     1     1
9537   41.959128 -87.720375     1     1
9538   41.959137 -87.710652     1     1
9539   41.959140 -87.709087     1     1
9540   41.959155 -87.708488     1     1
9541   41.959183 -87.717968     1     1
9542   41.959193 -87.719015     1     1
9543   41.959210 -87.710050     1     1
9544   41.959215 -87.726288     1     1
9545   41.959215 -87.709420     1     1
9521   41.959027 -87.712413     1     1
9519   41.959020 -87.730042     1     1
...          ...        ...   ...   ...
4797   41.894853 -87.763522     1     1
4798   41.894858 -87.714748     1     1
4774   41.894445 -87.756867     1     1
4773   41.894435 -87.723573     1     1
4772   41.894418 -87.715112     1     1
4771   41.894413 -87.710262     1     1
4748   41.894045 -87.756855     1     1
4749   41.894048 -87.714745     1     1
4750   41.894068 -87.715078     1     1
4751   41.894077 -87.766728     1     1
4752   41.894102 -87.721233     1     1
4753   41.894108 -87.715678     1     1
4754   41.894118 -87.716298     1     1
4755   41.894122 -87.722377     1     1
4756   41.894133 -87.707798     1     1
4757   41.894140 -87.724800     1     1
4758   41.894157 -87.754473     1     1
4759   41.894160 -87.767937     1     1
4760   41.894223 -87.711547     1     1
4761   41.894267 -87.759332     1     1
4762   41.894270 -87.709052     1     1
4763   41.894273 -87.718747     1     1
4764   41.894282 -87.748178     1     1
4765   41.894322 -87.713947     1     1
4766   41.894333 -87.712617     1     1
4767   41.894337 -87.714500     1     1
4768   41.894343 -87.760688     1     1
4769   41.894380 -87.772148     1     1
4770   41.894402 -87.704128     1     1
14293  42.395983 -88.095757     1     1

[14294 rows x 4 columns]

In [11]:
order_groups.loc[11499, :]['Latitude']

41.9839166666667

In [12]:
lat_long_count = spray.drop('Time', axis=1).groupby(['Latitude', 'Longitude']).agg(['count'])

lat_long_count.columns

lat_long_count = pd.Series(lat_long_count.values.reshape(1,-1)[0], lat_long_count.index)

lat_long_count.sort_values(ascending=False).head()

Latitude   Longitude 
41.986460  -87.794225    541
41.983917  -87.793088      2
41.894413  -87.710262      1
41.894380  -87.772148      1
41.894343  -87.760688      1
dtype: int64

Through manual inspection, we have determined a spray entry which has been duplicated 541 times.  We are going to remove the duplicates from this list.

In [13]:
mask = (spray['Latitude'] == 41.986460) & (spray['Longitude'] == -87.794225)
spray[mask].shape

(541, 4)

In [14]:
spray[mask].index[1:]

Int64Index([ 490,  491,  492,  493,  494,  495,  496,  497,  498,  499,
            ...
            1020, 1021, 1022, 1023, 1024, 1025, 1026, 1027, 1028, 1029],
           dtype='int64', length=540)

In [15]:
spray.drop(spray[mask].index[1:], inplace=True)

In [16]:
lat_long_count = spray.drop('Time', axis=1).groupby(['Latitude', 'Longitude']).agg(['count'])

lat_long_count = pd.Series(lat_long_count.values.reshape(1,-1)[0], lat_long_count.index)

lat_long_count.sort_values(ascending=False).head()

Latitude   Longitude 
41.983917  -87.793088    2
42.395983  -88.095757    1
41.894157  -87.754473    1
41.894380  -87.772148    1
41.894343  -87.760688    1
dtype: int64

In [17]:
mask = (spray['Longitude'] == -87.7930883333333) & (spray['Latitude'] ==  41.9839166666667)
sum(mask)

2

In [18]:
spray[mask].index[1:]

Int64Index([485], dtype='int64')

In [19]:
spray.drop(spray[mask].index[1:], inplace=True)

In [21]:
lat_long_count = spray.drop('Time', axis=1).groupby(['Latitude', 'Longitude']).agg(['count'])

lat_long_count = pd.Series(lat_long_count.values.reshape(1,-1)[0], lat_long_count.index)

lat_long_count.sort_values(ascending=False).head()

Latitude   Longitude 
42.395983  -88.095757    1
41.894160  -87.767937    1
41.894402  -87.704128    1
41.894380  -87.772148    1
41.894343  -87.760688    1
dtype: int64

In [22]:
spray.drop('Time', axis=1, inplace=True)

In [23]:
spray.to_csv('Spray_Cleaned.csv')